upt_macports/upt_macports.py

```diff
@@ -278,3 +278,32 @@ def package_versions(self, name):
             sys.exit(f'The command "{cmd}" failed. '
                      'Please make sure you have MacPorts installed '
                      'and/or your PATH is set-up correctly.')
+
+    def cpan_version(self, vers):
```

Comment:
the function itself works as intentded, at least to the same extent as what's happening in MacPorts (more on that below). However, the code looks a bit "ugly" - not very Pythonic and it's probably a direct "translation" of the Perl procedure. Perhaps we can improve on that a little.

Reply:
Yeah, I'd love to know whether you "converted" this from Perl code, and if so, if you could give us a link to this code.

This is also the perfect patch for TDD: write all your test cases, then make sure your function works.

Reply:
It's a direct translation of `proc perl5_convert_version {vers} ` from [perl portgroup](https://github.com/macports/macports-ports/blob/master/_resources/port1.0/group/perl5-1.0.tcl). 

Yeah, right...I'll add the tests.

Reply:
Sorry I meant to say the Tcl procedure... the code for converting "upstream" versions to what ```port info --version <package>``` returns (and is supposed to follow the Perl conversion) is part of the perl5 PortGroup. The file is found in the ```macports-ports``` repository under ```_resources/port1.0/group/perl5-1.0.tcl```. 

The relevant code is:
```
# Convert a floating-point version to an equivalent dotted decimal one.
# If version is expressed as a perl v-string, strip the leading "v"
proc perl5_convert_version {vers} {
    if {[string index $vers 0] eq "v"} {
        set start 1
    } else {
        set start 0
    }
    set index [string first . $vers]
    set other_dot [string first . [string range $vers [expr {$index + 1}] end]]
    if {$index == -1 || $other_dot != -1} {
        return [string range $vers $start end]
    }
    set ret [string range $vers $start [expr {$index - 1}]]
    incr index
    set fractional [string range $vers $index end]
    set index 0
    while {$index < [string length $fractional] || $index < 6} {
        set sub [string range $fractional $index [expr {$index + 2}]]
        if {[string length $sub] < 3} {
            append sub [string repeat "0" [expr {3 - [string length $sub]}]]
        }
        append ret ".[scan $sub %u]"
        incr index 3
    }
    return $ret
```

Reply:
why does the function has ```self``` as an argument - it doesn't seem to use it... or?